### ISSUE TITLE
Create integration tests for SPM

### ISSUE THREAD

**Issue by yurishkuro** [jaegertracing/jaeger#5608]
In #5603 the bot is trying to upgrade a dependency of the example. We don't actually know if that upgrade is safe or not, it may also require argument changes.

The main issue is that this `docker-compose/monitor/docker-compose.yml` file is not being tested in the CI. SPM is a major functionality, it's good to have some basic smoke test that it's working, as well as to test that the corresponding docker compose file is valid. An example of such e2e test is `scripts/build-all-in-one-image.sh` where we validate that the resulting image is correctly serving the web UI.

Proposal for integration test:
  * [x] perform a fresh build of Jaeger code (don't load published images), to test the current branch
  * [x] bring up services via docker compose
  * [x] wait for all services to respond successfully to health checks
  * [x] for each service in microsim
    * [x] loop until at least 3 data non-zero data points are retrieved from the REST API
    * [x] or until the overall timeout (say 10min) is reached, in which case fail the test
    * [ ] we need to test that the response also contains the right labels, specifically to catch this issue: https://github.com/jaegertracing/jaeger/pull/5673

Additional improvements:
  * [ ] The prometheus config is set to scrape metrics every 15sec, but in the UI the data points come with granularity of 1min. That means the test will have to run for at least 3min before succeeding, an artificial delay. We should find a way to use shorter interval (it may be already supported by the query API)
  * [ ] Check not just `calls` metric, but also `errors`, at least for services that may have errors
    * [x] This might require improvement to microsim to generate errors. https://github.com/yurishkuro/microsim/pull/16

**Comment by EraKin575** [comment]
Hi, @yurishkuro ! I am new contributor to this project! I would love to take this up!

**Comment by kunal-511** [comment]
Hi, I have made some changes according to this issue's requirements,  I am ready to make a PR request. @yurishkuro Please assign me.

**Comment by yurishkuro** [comment]
@kunal-511 Feel free to open pr. We don't assign issues.

**Comment by arujjval** [comment]
Is this issue still open?

**Comment by yurishkuro** [comment]
yes

**Comment by akstron** [comment]
Hi @yurishkuro , can you help with some more details for this:

> The prometheus config is set to scrape metrics every 15sec, but in the UI the data points come with granularity of 1min. That means the test will have to run for at least 3min before succeeding, an artificial delay. We should find a way to use shorter interval (it may be already supported by the query API)

The test is configured to wait for 10 seconds if the desired number of metrics is not found, or probably I am missing something.

**Comment by yurishkuro** [comment]
Code link?

**Comment by akstron** [comment]
https://github.com/jaegertracing/jaeger/blob/9c1c3ceac3c3c02af0970cabeac8d91aa5b83529/scripts/spm-integration-test.sh#L141

**Comment by yurishkuro** [comment]
it doesn't wait 10sec, it waits `while [ $SECONDS -lt $end_time ]; do`

**Comment by akstron** [comment]
> it doesn't wait 10sec, it waits while [ $SECONDS -lt $end_time ]; do

Yes, I understand that. If I understand it correctly, the script calls the `/metric` api every 10 seconds, but the backend of `/metric` itself aggregates the metric from prometheus with a granularity of 1 min?

**Comment by yurishkuro** [comment]
it makes no difference how often the script calls the endpoint, it does not change the shape of the data. The parameters of the query do change the shape https://github.com/jaegertracing/jaeger/blob/9c1c3ceac3c3c02af0970cabeac8d91aa5b83529/scripts/spm-integration-test.sh#L81

**Comment by Don-Assamongkol1** [comment]
Hey there! Is this issue still open? Seems like it was addressed in https://github.com/jaegertracing/jaeger/pull/6991

**Comment by yurishkuro** [comment]
it's not fully addressed

**Comment by Don-Assamongkol1** [comment]
Oh whoops–- I see the incomplete bullet points in the description. 

@yurishkuro from the description it seems that you blocked the upgrade of the microservices simulator image from 0.2.0 -> 0.3.0 (https://github.com/jaegertracing/jaeger/pull/5603) because you wanted an E2E test to verify that it's working. 

But I see that on main, the simulator image is already at the latest tag (0.5.0): https://github.com/jaegertracing/jaeger/blob/96588226fc0e8f3422e4725ebcb400bee022a15a/docker-compose/monitor/docker-compose.yml#L22. 

Is the issue still open because you'd like the other points to be tested? I.e. 
> we need to test that the response also contains the right labels, specifically to catch this issue: https://github.com/jaegertracing/jaeger/pull/5673

If so, I'd be interested in giving this issue a try if no one else is actively working on it.

**Comment by yurishkuro** [comment]
yes, it's open for the remaining items in the checklist. Feel free to take it.

**Comment by Don-Assamongkol1** [comment]
@yurishkuro for the item
> Check not just `calls` metric, but also `errors`, at least for services that may have errors

I think the test already tests this endpoint: https://github.com/jaegertracing/jaeger/blob/e91d0282d5f815ab90b7e01acf8bda5891cccfa7/scripts/e2e/spm.sh#L165-L171. 

What additional behavior would you want tested for the errors endpoint? 

I think the test I added to check labels for the calls endpoint also suffices as a test to check labels for the errors endpoint. If I understand correctly, the `/metrics/errors` endpoint returns the error rate, but the spans made in that query should be a subset of the spans made in the `/metrics/calls` endpoint. So I think that testing labels in the call endpoint also verifies that labels exist in the errors endpoint. 

Eg the prometheus query for errors seems to be just a stricter filter of the query for the calls endpoint: 
https://github.com/jaegertracing/jaeger/blob/e91d0282d5f815ab90b7e01acf8bda5891cccfa7/internal/storage/metricstore/prometheus/metricstore/reader.go#L183

https://github.com/jaegertracing/jaeger/blob/e91d0282d5f815ab90b7e01acf8bda5891cccfa7/internal/storage/metricstore/prometheus/metricstore/reader.go#L217

**Comment by Don-Assamongkol1** [comment]
Also @yurishkuro could you explain what you mean UI data points have a 1min granularity ==> this test has to run 3 min before succeeding? 

My understanding so far: 
1. So for the Monitor UI having 1 min granularity, I take that to mean the Jaeger UI queries for metrics spaced 1 min apart. That is, when I move my cursor on the Monitor tab to the right, the next data point showed would be for 12:13:10
<img width="643" height="469" alt="Image" src="https://github.com/user-attachments/assets/903db023-3317-4772-977a-17d3cb95fd8e" />

2. I guess taking a step back, I'm confused why we have to wait at all in the first place. The microsim app should be producing requests [every 500ms](https://github.com/jaegertracing/jaeger/blob/e91d0282d5f815ab90b7e01acf8bda5891cccfa7/docker-compose/monitor/docker-compose.yml#L23) in the test. These requests flow through the microsim architecture generating traces which the jaeger exporter makes available for Prometheus to scrape, which it does every 15 seconds. Assuming that the jaeger exporter isn't slow, I'd expect that metrics should be available to scrape very early on. But instead I see "Metrics not available yet" for more than a minute (the logs are separated by 10s): 

```
Waiting for services to be up and running...
Checking health of service: Prometheus at http://localhost:9090/query
Waiting for Prometheus to be healthy...
✅ Prometheus is healthy
Checking health of service: Jaeger at http://localhost:16686
✅ Jaeger is healthy
Processing service: driver
⏳ Metrics not available yet (no metrics array)
⏳ Metrics not available yet (no metrics array)
⏳ Metrics not available yet (no metrics array)
⏳ Metrics not available yet (no metrics array)
⏳ Metrics not available yet (no metrics array)
⏳ Metrics not available yet (no metrics array)
⏳ Metrics not available yet (no metrics array)
⏳ Metrics not available yet (no metrics array)
⏳ Metrics not available yet (no metrics array)
```

**Comment by yurishkuro** [comment]
> The microsim app should be producing requests every 500ms in the test. These requests flow through the microsim architecture generating traces which the jaeger exporter makes available for Prometheus to scrape, which it does every 15 seconds. Assuming that the jaeger exporter isn't slow, I'd expect that metrics should be available to scrape very early on. But instead I see "Metrics not available yet" for more than a minute

That's precisely the problem. When metrics query service receives a request and proxies it to Prometheus it specifies at which granularity Prometheus should return the results, and somewhere that is set to 1min. So even though metrics are scraped every 15sec when we ask for 1min granularity we are not seeing data until 1min passed. For the integration tests we could speed this up. And we could also configure Prometheus to scrape even more frequently, like every 5sec.

**Comment by Don-Assamongkol1** [comment]
> > The microsim app should be producing requests every 500ms in the test. These requests flow through the microsim architecture generating traces which the jaeger exporter makes available for Prometheus to scrape, which it does every 15 seconds. Assuming that the jaeger exporter isn't slow, I'd expect that metrics should be available to scrape very early on. But instead I see "Metrics not available yet" for more than a minute
> 
> That's precisely the problem. When metrics query service receives a request and proxies it to Prometheus it specifies at which granularity Prometheus should return the results, and somewhere that is set to 1min. So even though metrics are scraped every 15sec when we ask for 1min granularity we are not seeing data until 1min passed. For the integration tests we could speed this up. And we could also configure Prometheus to scrape even more frequently, like every 5sec.

looking

**Comment by Don-Assamongkol1** [comment]
@yurishkuro I'm trying to get the Integration Test to go faster, but I think what I'm finding doesn't align with your hint above. Would you be able to let me know if I'm potentially going down a wrong rabbit hole? 

If I understand correctly, what we pass to the Prometheus query can be specified by the REST call we make. Setting the step to 15sec means Prometheus should calculate the value for the metric every 15 secs; had it been one minute, it would make sense that Prometheus only returns data after 1 minute because that's when we configured it to calculate the metric. 

However, the `step` is [already set to 15secs](https://github.com/jaegertracing/jaeger/blob/0bea7f0057668d1a8e96ae1367ae32dfa3f01bed/scripts/e2e/spm.sh#L227) so I don't think this is the issue. 

Looking at the logs, it seems that Jaeger (I'm not sure which component in Jaeger exactly) isn't serving data up to Prometheus until the one minute mark. 

Here's an excerpt from me running `spm.sh` locally: 
```
[2026-01-09 15:35:59] Processing service: driver
[2026-01-09 16:09:03] ⏳ Metrics not available yet (no metrics array)
[2026-01-09 16:09:13] ⏳ Metrics not available yet (no metrics array)
[2026-01-09 16:09:23] ⏳ Metrics not available yet (no metrics array)
[2026-01-09 16:09:33] ⏳ Metrics not available yet (no metrics array)
[2026-01-09 16:09:43] ⏳ Metrics not available yet (no metrics array)
[2026-01-09 16:09:53] ⏳ Metrics not available yet (no metrics array)
[2026-01-09 16:10:03] ⏳ Metrics not available yet (no metrics array)
[2026-01-09 16:10:13] ⏳ Metrics data points found: 1 zero, 0 non-zero
[2026-01-09 16:10:13] ⏳ Expecting at least 4 non-zero data points
[2026-01-09 16:10:24] ⏳ Metrics data points found: 1 zero, 0 non-zero
[2026-01-09 16:10:24] ⏳ Expecting at least 4 non-zero data points
[2026-01-09 16:10:34] ⏳ Metrics data points found: 2 zero, 0 non-zero
[2026-01-09 16:10:34] ⏳ Expecting at least 4 non-zero data points
[2026-01-09 16:10:44] ⏳ Metrics data points found: 3 zero, 0 non-zero
[2026-01-09 16:10:44] ⏳ Expecting at least 4 non-zero data points
[2026-01-09 16:10:54] ⏳ Metrics data points found: 3 zero, 0 non-zero
[2026-01-09 16:10:54] ⏳ Expecting at least 4 non-zero data points
[2026-01-09 16:11:04] ⏳ Metrics data points found: 4 zero, 0 non-zero
[2026-01-09 16:11:04] ⏳ Expecting at least 4 non-zero data points
[2026-01-09 16:11:14] ⏳ Metrics data points found: 4 zero, 1 non-zero
[2026-01-09 16:11:14] ⏳ Expecting at least 4 non-zero data points
[2026-01-09 16:11:24] ⏳ Metrics data points found: 4 zero, 1 non-zero
[2026-01-09 16:11:24] ⏳ Expecting at least 4 non-zero data points
```

You can see that the first time stamp where the test script can find metric data points (16:10:13) comes after the first time that prometheus actually scraped data: (the following link is me copying logs out of the jaeger container)

```
2026-01-09T16:10:06.515Z	debug	prometheusexporter@v0.142.0/collector.go:628	metric served: Desc{fqName: "traces_span_metrics_calls_total", help: "", constLabels: {}, variableLabels: {service_name,span_name,span_kind,status_code,otel_scope_name,otel_scope_version,otel_scope_schema_url,job}}	
```
from https://github.com/Don-Assamongkol1/jaeger/blob/5a725647dc75299e6fcb7f2dfcc22b67e5e1cc46/docker_logs.log#L208C77-L208C264

So I guess my question becomes: why does Jaeger not have metrics to serve up until 1 minute? 

My mental model is that:
- microsim sends traces every 500ms via OTLP instrumentation
- these get collected to Jaeger
- But how do these get exposed to Prometheus? Is it that the moment it comes into Jaeger, metrics for these spans are immediately available for Prometheus to scrape? Seems not. It seems that this is where the 1 minute wait happens; Prometheus is scraping every 5 seconds or so, but the metrics only become available after 1 minute. 

I'm not sure which component is making Jaeger wait to serve metrics. I assume that this is in place because of some batch processing for Jaeger to send metrics out, and I'm guessing for the integration test that this batch interval can be tuned so that metrics become available more quickly. I.e we can only set the Prometheus scrape interval to be as low as the interval that Jaeger can expose metrics. 

Would you be able to give guidance on (a) if this is correct and (b) how Jaeger serves up metrics? (Or is it OTLP that's doing the caching?)

**Comment by yurishkuro** [comment]
metrics are produced from spans by spanmetricsconnector which is an exporter from the traces pipeline and a receiver in the metrics pipeline. Then the metrics pipeline exposes those metrics via prometheus exporter on a port for prometheus to scrape (I actually think we could've used a push method via `prometheusremotewrite` exporter).

So it sounds like there may be some delay between batch processor and spanmetricsconnector before the metrics become available for scraping.
